### PR TITLE
[video_player_avfoundation] send video load failure even when eventsink was initialized late

### PR DESCRIPTION
**Pull request by misos1**
Event `initialized` is sent from `onListenWithArguments` by calling `setupEventSinkIfReadyToPlay` if event sink was `nil` during `observeValueForKeyPath`. This change also sends failure in a similar way. Adds more details to the error message returned by `VideoPlayerController.initialize()`.

- https://github.com/flutter/flutter/issues/151475
- https://github.com/flutter/flutter/issues/147707
- https://github.com/flutter/flutter/issues/56665 - now error message may look like: `Failed to load video: Operation Stopped: The server is not correctly configured.: The operation couldn’t be completed. (CoreMediaErrorDomain error -12939 - byte range and no content length - error code is 200)`

Tests `testSeekToWhilePausedStartsDisplayLinkTemporarily` and `testSeekToWhilePlayingDoesNotStopDisplayLink` are failing on the main branch on ios. Seems position is correctly set to 1234 in `seekTo` completion handler but right after `[mockDisplayLink setRunning:YES]` is set again to 0 and after some time back to 1234 so those two tests sometimes fail (seems more often when running all tests).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:404]: it looks like `sendFailedToLoadVideoEvent` is only called from background queue. Can this function simply be: 
```
- (void)sendFailedToLoadVideoEventOnMainQueue:(NSError *)error {
  dispatch_async(dispatch_get_main_queue(), ^{
    // all logic here
  }
}
```

Comment (misos1) [reply packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:404]: It is also called from `setupEventSinkIfReadyToPlay` and `observeValueForKeyPath`.

**Comment by hellohuanlin** [comment]
I have a hard time understanding this in the PR description: 

> Sends also failure similarly as is sent event initialized from onListenWithArguments through setupEventSinkIfReadyToPlay if event sink was not initialized during observeValueForKeyPath. 

Could you rephrase a bit? That'd be helpful for a review. Thanks!

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:348]: nit: `sendFailedToLoadVideoEventIfNeeded` since you moved the check inside

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:421]: i dont think we need this comment

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:435]: Can you separate these 4 lines into multiple sentences? I have a hard time understanding this.

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:464]: can you also specify which thread this is on

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (misos1) [reply packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:348]: I would not say "if needed" sounds right here. Meaning of that check is rather like "if can" or "if possible".

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (misos1) [reply packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:464]: On some AVFoundation background queue (this is also mentioned in a comment a few lines above):

https://developer.apple.com/documentation/avfoundation/avasynchronouskeyvalueloading/1387321-loadvaluesasynchronouslyforkeys

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (hellohuanlin) [reply packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:414]: can remove this link.

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:432]: can we move this to the caller, before `setupEventSinkIfReadyToPlay` is called?

Comment (hellohuanlin) [reply packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:432]: or can you update the function name to `setupEventSinkIfReadyToPlayOrReportFailure`

Comment (misos1) [reply packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:432]: There is also 4th place where this function is called masked by `_cmd` here `self performSelector:_cmd`, I would rather not put that check outside of that function.

Comment (misos1) [reply packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:432]: Actually that function does not set up `_eventsink`, I would say it is set up by `onListenWithArguments`. This function just sends an event through that event sink.

Comment (misos1) [reply packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:431]: On the other hand if status becomes `failed` on that second call it should be already caught in the key observer where the event sink is not nil and so it would send an error from there.

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:446]: is this code the key change that fixed the bug?

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:616]: by "similar case", do you mean AVPlayerItemStatusReadyToPlay fires before _eventSink is set? Can you be more explicit here in the comment? (Sorry I have some trouble understanding this bug fix)

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (misos1) [reply packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:446]: Seems no, I think it is that change in `onListenWithArguments`. But without that change it seems also this change works as I now tested it with "Code sample 1" from https://github.com/flutter/flutter/issues/151475.

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (misos1) [reply packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:616]: Yes I was following up with that comment above. But this time with `AVPlayerItemStatusFailed` instead of `AVPlayerItemStatusReadyToPlay`. If anything fires before is event sink set then it could not be sent. Would that comment be better if I added "in similar case as above"?

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (hellohuanlin) [inline packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:613]: So this is the part that actually fix the bug? Can you add a comment explaining how we ended up with Failed state here?

**Review by hellohuanlin — COMMENTED** [review]
Sorry for the delay.

**Comment by stuartmorgan** [comment]
@hellohuanlin Please add tarrinneal (the ecosystem-level CODEOWNER) as the secondary reviewer once this has passed iOS review; the iOS team review is primary since this is internal to the iOS implementation.

**Review by hellohuanlin — APPROVED** [review]
LGTM after the nit on comment

**Review by tarrinneal — APPROVED** [review]

File: packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m
Comment (misos1) [reply packages/video_player/video_player_avfoundation/darwin/video_player_avfoundation/Sources/video_player_avfoundation/FVPVideoPlayerPlugin.m:613]: Sorry I did not manage to proceed with this. Yes when `AVPlayerItemStatusFailed` fires before `_eventSink` is set. `AVPlayerItem` is "started" to do its things and can asynchronously become `ReadyToPlay` or `Failed` before the dart side starts to listen where `onListenWithArguments` is called. My comment here follows up that comment before which tries to explain this but that previous fix connected with that comment omitted `Failed` state as possible state here which this line fixes (if player item can end up with success here it can also end up with failure, so that previous comment should explain it).

**Comment by stuartmorgan** [comment]
Thanks, I closed one, and left a comment in another that is likely fixed but needs re-testing (unless you've already tested that one? If so we can close it). For the third it's not clear to me whether it's fully resolved.

**Pull request merged**
